src/protocol/protocol_set.rs

```diff
@@ -274,27 +277,36 @@ impl ProtocolSet {
         protocol: ProtocolName,
         direction: Direction,
         substream: Substream,
-    ) -> crate::Result<()> {
+    ) -> Result<(), SubstreamError> {
         tracing::debug!(target: LOG_TARGET, %protocol, ?peer, ?direction, "substream opened");
 
         let (protocol, fallback) = match self.fallback_names.get(&protocol) {
             Some(main_protocol) => (main_protocol.clone(), Some(protocol)),
             None => (protocol, None),
         };
 
-        self.protocols
-            .get_mut(&protocol)
-            .ok_or(Error::ProtocolNotSupported(protocol.to_string()))?
+        let Some(protocol_context) = self.protocols.get(&protocol) else {
```

Comment:
nit: Actually, I liked how it was implemented rust-way with `ok_or()`. `let .. else` feels somewhat C++-ish.

May be add some `#[from]` macros to the types and keep `.ok_or()?`?